### PR TITLE
Upgrade GPU CI to PyTorch 1.13

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #15555

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .azure/gpu-tests-pytorch.yml
Comment (awaelchli) [inline .azure/gpu-tests-pytorch.yml:47]: how come we had 11.6.1 here while the docker file had 11.3.1?

Comment (carmocca) [reply .azure/gpu-tests-pytorch.yml:47]: Because that's the dockerfile default value. The images are constructed from `.github/workflows/release-docker.yml` which you need to update as I did in https://github.com/Lightning-AI/lightning/pull/15347/files#diff-aa38a2216a456b6851247d357e99f30026306296d43bb3c1f843a6614524ef77R22

**Review by carmocca — COMMENTED** [review]
LGTM. Fine updating this first separately to the rest of CI.

Will approve once CI is green and it's actually using 1.13

File: dockers/base-cuda/Dockerfile
Comment (Borda) [inline dockers/base-cuda/Dockerfile:22]: I think there is some issue with Bagua, https://github.com/Lightning-AI/lightning/pull/15403#issuecomment-1297402599

**Comment by awaelchli** [comment]
How naive of me to think I could just update a few variables from 1.12 to 1.13. I'm just digging my own grave here if I continue.  Maybe this isn't as important as I thought. Sorry for the noise.

**Comment by carmocca** [comment]
I'm hijacking this

**Comment by carmocca** [comment]
CI stuff is ready, now blocked by a failing FSDP test: https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=115069&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=8e4ceb7c-ceed-5ee0-b6fc-c9023e41cb74&l=2191 cc @awaelchli

**Comment by awaelchli** [comment]
I can take a look too this afternoon!

**Comment by Borda** [comment]
> I can take a look too this afternoon!

as spoke, let's run PT 1.12 with all strategies and then PT 1.13 just for dependencies available at this time as running with 1.13 is important for us for testing the recent FSDP feature...

File: .azure/gpu-tests-pytorch.yml
Comment (carmocca) [inline .azure/gpu-tests-pytorch.yml:107]:
```suggestion
```

**Comment by awaelchli** [comment]
@carmocca @Borda It seems that PT 1.13 is setting some env vars that trigger our leakage assertions. These are the names:
```
KMP_INIT_AT_FORK  
KMP_DUPLICATE_LIB_OK
```
Shall we add them to the allowlist? There seem to be several tests that experience this leakage.

**Comment by carmocca** [comment]
Yes

**Review by carmocca — APPROVED** [review]
LGTM once fsdp tests are fixed

**Comment by awaelchli** [comment]
@carmocca FYI, the reshard_after_forward attribute that caused the attribute error was removed in PyTorch 1.13: https://github.com/pytorch/pytorch/pull/83665
I need to study the the commit in PyTorch to figure out what the alternative is to testing this. For now, I have removed the assertions from the tests in case you want to merge the PR. I can do this investigation as part of #14898

File: dockers/base-cuda/Dockerfile
Comment (Borda) [inline dockers/base-cuda/Dockerfile]: this could be a problem as we are already using docker with cuda

Comment (awaelchli) [reply dockers/base-cuda/Dockerfile]: I have absolutely no idea what I'm doing btw

File: .github/workflows/ci-pytorch-dockers.yml
Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:5]:
```suggestion
    branches: [master, "release/*"]
```

File: .github/workflows/ci-pytorch-dockers.yml
Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:27]:
```suggestion
  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
```

File: dockers/base-cuda/Dockerfile
Comment (awaelchli) [inline dockers/base-cuda/Dockerfile:166]: In the job above, we were compiling horovod from source (latest version). Here in the strategy.txt file however, the version was set to <0.25.1 which triggered a re-install, without cuda support. That's why we saw the horovod tests failing.

File: .github/checkgroup.yml
Comment (awaelchli) [inline .github/checkgroup.yml:138]: @carmocca what about these todos that you added?
what is the difference between build-cuda and build-pl?

File: .azure/gpu-benchmark.yml
Comment (awaelchli) [inline .azure/gpu-benchmark.yml:49]: The cuda_ver variable here did not have a value. It caused torchvision to be installed with +cu117 but we needed cu116 to match the torch version. Hardcoding it here since I didn't know what else to do

File: .github/checkgroup.yml
Comment (awaelchli) [inline .github/checkgroup.yml:131]:
```suggestion
      - "build-cuda (3.9, 1.13, 11.6.1)"
```

File: .github/checkgroup.yml
Comment (carmocca) [reply .github/checkgroup.yml:138]: The latter is for cpu jobs and should be left for Jirka's PR

File: .azure/gpu-benchmark.yml
Comment (carmocca) [reply .azure/gpu-benchmark.yml:49]: You can copy it from the other workflows. Removing it was probably an accident

Comment (awaelchli) [reply .azure/gpu-benchmark.yml]: that worked!

File: .github/workflows/ci-pytorch-dockers.yml
Comment (awaelchli) [inline .github/workflows/ci-pytorch-dockers.yml:25]: Before merge:
```suggestion
  PUSH_TO_HUB: ${{ github.event_name == 'schedule' }}
```

Comment (carmocca) [inline .github/workflows/ci-pytorch-dockers.yml:25]: Reminder to revert this once it's pushed

**Review by carmocca — APPROVED** [review]

File: .github/workflows/release-docker.yml
Comment (awaelchli) [inline .github/workflows/release-docker.yml:5]: Before merge:
```suggestion
    branches: [master, "release/*"]
```

File: .azure/gpu-benchmark.yml
Comment (Borda) [inline .azure/gpu-benchmark.yml:70]:
```suggestion
        python .actions/assistant.py requirements_prune_pkgs --packages [horovod,bagua,colossalai] --req_files [requirements/pytorch/strategies.txt]
```

File: .github/workflows/ci-pytorch-dockers.yml
Comment (Borda) [inline .github/workflows/ci-pytorch-dockers.yml:40]:
```suggestion
          - {python_version: "3.9", pytorch_version: "1.13", cuda_version: "11.6.1"}
```
I think this shall be added, right?

File: .azure/gpu-tests-pytorch.yml
Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:44]: but the name for export is based inputs not defaults

File: dockers/base-cuda/Dockerfile
Comment (Borda) [inline dockers/base-cuda/Dockerfile:165]: if we copy assistant we can simply use:
```
python .actions/assistant.py requirements_prune_pkgs --packages [horovod,colossalai] --req_files [requirements/pytorch/strategies.txt]
```

**Review by Borda — APPROVED** [review]

File: dockers/base-cuda/Dockerfile
Comment (carmocca) [reply dockers/base-cuda/Dockerfile:165]: I suggest you do this change for all workflows in a separate PR. This doesn't need to be done here

Comment (Borda) [reply dockers/base-cuda/Dockerfile:165]: yeah, that is why I did not push it :)

**Pull request merged**
